onnxruntime/core/providers/cuda/reduction/reduction_functions.cu

```diff
@@ -0,0 +1,401 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include <algorithm>
+#include <cuda.h>
+#include <cuda_fp16.h>
+#include "core/providers/cuda/cu_inc/common.cuh"
+#include "core/providers/cuda/atomic/common.cuh"
+#include "reduction_functions.h"
+#include "reduction_utils.cuh"
+
+#define NUM_ELEMENTS_PER_THREAD 4
+#define NUM_THREADS_PER_WARP 32
+#define NUM_WARPS_PER_BLOCK 8
```

Comment:
Is this equivalent to 'constexpr int CUDA_WARP_SIZE = 32;' in softmax_impl.cuh?
If so, should this sort of value be defined in a more common location?

Reply:
Yes. It's equivalent to that. This attribute defines the number of reduced elements per warp.